### PR TITLE
Fix for SI-6662, macro failing too early.

### PR DESCRIPTION
**Pull request by paulp**
Don't give up before you try tupling. Who knows what
someone might be doing with a Unit.

### CODE REVIEW

**Comment by paulp** [comment]
Review by @xeno-by

**Comment by xeno-by** [comment]
I wonder if `checkNotMacro` is needed here. https://github.com/scalamacros/kepler/blob/2e0cbe0aa276720ceaf54a3448f7e04558e255b8/src/compiler/scala/tools/nsc/typechecker/Typers.scala#L3102

**Comment by paulp** [comment]
Doesn't look necessary, I agree.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -3037,7 +3037,7 @@ trait Typers extends Modes with Adaptations with Tags {
 
             def checkNotMacro() = {
               if (fun.symbol != null && fun.symbol.filter(sym => sym != null && sym.isTermMacro && !sym.isErroneous) != NoSymbol)
-                duplErrorTree(NamedAndDefaultArgumentsNotSupportedForMacros(tree, fun))
+                tryTupleApply getOrElse duplErrorTree(NamedAndDefaultArgumentsNotSupportedForMacros(tree, fun))
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3040]: From what I understood, if `tryTupleApply` returns true, then the subsequent logic in `tryNamesDefaults` will choose tuple application and won't go for named or default argument applications. Right?

**Comment by retronym** [comment]
Could you retarget this at `2.10.x`? I'm already struggling to decide what needs backporting in `git log master ^2.10.x  --oneline --graph`.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 1 above)
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3040]: Well, actually, after looking into Typers more closely, it seems that default arguments take precedence over tuple applications, so augmenting `checkNotMacro` might lead to unexpected results.

Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3040]: Who knows, in this all-side-effect universe. If we'd write functions which returned things questions like this wouldn't even come up.

**Comment by paulp** [comment]
paulp lacks the necessary patience to navigate the hedge maze. Everyone has my permission to move my patches around though.

**Comment by xeno-by** [comment]
All in all, lgtm. 

I think we can accept this not totally bulletproof solution in favor of resolving an obvious and annoying bug. Named/default arguments and macros aren't supported anyways, so when their time comes, I will just remove `checkNotMacro` and the entire point of the dicsussion above will become moot.

**Comment by xeno-by** [comment]
https://github.com/scala/scala/pull/1618

**Pull request closed without merging**